lib/src/sources/dmi_source.cc

```diff
@@ -0,0 +1,207 @@
+        }
+
+        boost::trim(value);
+
+        // Replace any non-printable ASCII characters with '.'
+        // This mimics the behavior of dmidecode
+        for (auto& c : value) {
+            if (c < 32 || c == 127) {
+                c = '.';
+            }
+        }
+        return value;
+    }
+
+    void dmi::parse_dmidecode_line(string& line, int& dmi_type)
+    {
+        static const boost::regex dmi_section_pattern{"^Handle 0x.{4}, DMI type (\\d{1,3})"};
+
+        // Stores the relevant sections; this is in order based on DMI type ID
+        // Ensure there's a trailing semicolon on each entry and keep in sync with the switch statement below
+        static const vector<vector<string>> sections{
+            {   // BIOS (0)
+                "vendor:",
+            },
+            {   // System (1)
+                "manufacturer:",
+                "product name:",
+            },
+            {   // Base Board (2)
+                "manufacturer:",
+                "product name:",
+            }
+        };
+
+        // Check for a section header
+        if (re_search(line, dmi_section_pattern, &dmi_type)) {
+            return;
+        }
+
+        // Check that we're in a relevant section
+        if (dmi_type < 0 || static_cast<size_t>(dmi_type) >= sections.size()) {
+            return;
+        }
+
+        // Trim leading whitespace
+        boost::trim_left(line);
+
+        // Find a matching header
+        auto const& headers = sections[dmi_type];
+        auto it = find_if(headers.begin(), headers.end(), [&](string const& header) {
+            return boost::istarts_with(line, header);
+        });
+        if (it == headers.end()) {
+            return;
+        }
+
+        // Get the value and trim it
+        string value{line.substr(it->size())};
+        boost::trim(value);
+
+        // Calculate the index into the header vector
+        auto index = it - headers.begin();
+
+        // Assign to the appropriate member
+        string* member{nullptr};
+        switch (dmi_type) {
+            case 0: {  // BIOS information
+                if (index == 0) {
+                    member = &data_->bios_vendor;
+                }
+                break;
+            }
+
+            case 1: {  // System information
+                if (index == 0) {
+                    member = &data_->manufacturer;
+                } else if (index == 1 || index == 2) {
+                    member = &data_->product_name;
+                }
+                break;
+            }
+
+            case 2: {  // Base board information
+                if (index == 0) {
+                    member = &data_->board_manufacturer;
+                } else if (index == 1 || index == 2) {
+                    member = &data_->board_product_name;
+                }
+                break;
+            }
+
+            default:
+                break;
+        }
+
+        if (member) {
+            *member = std::move(value);
```

Comment:
That's right - this part comes from facter's [linux dmi resolver](https://github.com/puppetlabs/facter/blob/11288d2a33caec577fe696bd02e4bed12e5ac12d/lib/src/facts/linux/dmi_resolver.cc#L116)